### PR TITLE
docs: add text tutorial

### LINKED ISSUE #335: Create text based search example for documentation
Include quality improvements on the chosen dataset

stop quality improvements after 26th of January

### PR DESCRIPTION
**Pull request by gmastrapas**
Closes #335 

Example of using finetuner for text classification.

### CODE REVIEW

**Review by azayz — CHANGES_REQUESTED** [review]
- Add comment to code
- Last long code block should be dropdown (check my tutorial) 
Screenshot from 2022-01-26 16-52-33

File: docs/get-started/clinc150/index.md [hunk 1]
```diff
@@ -0,0 +1,842 @@
...
+        padding=True,
+        return_tensors='pt',
+    )
+```
+
+## Pre-trained model performance
+
+Before proceeding with the fine-tuning, we should evaluate the performance of the
+pre-trained model, so that after fine-tuning we can evaluate again and compare the results.
+To evaluate the model, we can use the built-in `Evaluator` component that allows us to
```
Comment (makram93) [inline docs/get-started/clinc150/index.md:190]:
```suggestion
To evaluate the model, we can use the built-in `Evaluator` component of the `finetuner` that allows us to
```

File: docs/get-started/clinc150/index.md [hunk 2]
```diff
@@ -0,0 +1,842 @@
...
+
+Out of the box, the pre-trained model has a solid performance on our dataset. This means
+that pre-training enables the model to learn good representations that map semantically
+similar texts nearby in the embedding space. Let's see if we can improve using `finetuner`.
+
+
+## Fine-tuning
+
+We will finetune the model for 6 epochs using a batch size of 256. We are using a learning
+rate of 1e-4, with the AdamW optimizer and a linear learning rate schedule with warmup.
```
Comment (makram93) [inline docs/get-started/clinc150/index.md:236]:
```suggestion
rate of 1e-4, with the AdamW optimizer and a linear learning rate scheduler with warmup.
```

File: docs/get-started/clinc150/index.md [hunk 3]
```diff
@@ -0,0 +1,842 @@
...
+For the training objective, we are using the `TripletLoss` in conjunction with the
+`TripletEasyHardMiner` with easy positive and hard negative strategies and a margin of `0.4`.
+
+Finally we make use of the various callbacks provided by `finetuner`, to inject
+functionalities in the training loop. Specifically, we use the `EvaluationCallback`
+so that metrics are computed on the val set after each epoch, the `EarlyStopping`
+callback which monitors the average precision to trigger early stopping if the metric stops
+increasing, the `BestModelCheckpoint` that saves the best performing model (in terms of
+average precision) every epoch and finally the `WandBLogger` callback that logs our
+training information using Weights and Biases.
```
Comment (makram93) [inline docs/get-started/clinc150/index.md:251]:
```suggestion
training information using [Weights and Biases](https://wandb.ai/site).
```

File: docs/get-started/clinc150/index.md [hunk 4]
```diff
@@ -0,0 +1,842 @@
...
+    device=DEVICE,
+    batch_size=BATCH_SIZE,
+    collate_fn=collate_fn,
+)
+```
+
+Let's now use this method to run some test cases.
+
+```python
+utterance = 'Where do you thing I should travel to this Christmas?'
```
Comment (makram93) [inline docs/get-started/clinc150/index.md:460]: Is it intentional? 
```suggestion
utterance = 'Where do you think I should travel to this Christmas?'
```

**Review by makram93 — COMMENTED** [review]
Some minor comments. Pretty interesting tutorial 👍

File: docs/get-started/clinc150/index.md (hunk 4 above)
Comment (gmastrapas) [reply docs/get-started/clinc150/index.md:460]: Nope 😛  Now I need to re-run the tutorial

**Review by makram93 — APPROVED** [review]
LGTM!

File: docs/get-started/clinc150/index.md [hunk 5]
```diff
@@ -0,0 +1,843 @@
...
+We will use the CLINC150 dataset as the base of our experiment. It is a dataset of
+utterance-intent pairs and is commonly used for evaluating intent models. It comes
+in train, val and test splits and contains 150 intents from various chatbot domains.
+For more info on the CLINC150 dataset, check out the
+[dataset repo](https://github.com/clinc/oos-eval).
+
+Firstly, let's clone the dataset locally:
+```
+git clone https://github.com/clinc/oos-eval.git
+```
```
Comment (makram93) [inline docs/get-started/clinc150/index.md:42]:
```suggestion
```bash
git clone https://github.com/clinc/oos-eval.git
```
```

File: docs/get-started/clinc150/index.md [hunk 6]
```diff
@@ -0,0 +1,843 @@
...
+
+As described above, we will use `finetuner` to finetune an embedding model in order
+to bring representations of the same intent, closer in the embedding space. For that
+we will use the `transformers` library to define a transformer-based embedding model.
+We will load a pre-trained transformer as our starting point, the `paraphrase-MiniLM-L6-v2`
+model from `sentence-transformers`. Check out the links below for more information on
+this specific model as well as sentence transformers in general:
+
+* https://huggingface.co/sentence-transformers/paraphrase-MiniLM-L6-v2
+* https://www.sbert.net/index.html
```
Comment (azayz) [inline docs/get-started/clinc150/index.md:133]: should be clickable

File: docs/get-started/clinc150/index.md [hunk 7]
```diff
@@ -0,0 +1,843 @@
...
+To evaluate the model, we can use the built-in `Evaluator` component of `finetuner` that
+allows us to compute information retrieval metrics. We will evaluate on the test split
+of our dataset:
+
+```python
+from finetuner.tuner.evaluation import Evaluator
+
+BATCH_SIZE = 256
+NUM_WORKERS = 8
+DEVICE = 'gpu' if torch.cuda.is_available() else 'cpu'
```
Comment (azayz) [inline docs/get-started/clinc150/index.md:199]: should be 'cuda' and not 'gpu'

**Review by azayz — CHANGES_REQUESTED** [review]

File: docs/get-started/clinc150/index.md [hunk 8]
```diff
@@ -0,0 +1,843 @@
...
+Specifically we will tune a transformer model on an Intent Classification task. Intent
+classification is the problem where we try to predict the user intent from a user utterance.
+It is a common step in chatbots and Conversational AI, where after the user speech has been
+decoded to text, we try to represent the meaning of the text symbolically by predicting
+intents and semantic entities. For example:
+
+* I want to book a flight - intent: `book-flight`
+* What is the weather forecast for tomorrow? - intent: `get-weather`
+
+The intent classification task is usually formulated as text classification i.e. we build
```
Comment (tadejsv) [inline docs/get-started/clinc150/index.md:21]: Split this paragraph into two

File: docs/get-started/clinc150/index.md [hunk 9]
```diff
@@ -0,0 +1,843 @@
...
+
+We will use the CLINC150 dataset as the base of our experiment. It is a dataset of
+utterance-intent pairs and is commonly used for evaluating intent models. It comes
+in train, val and test splits and contains 150 intents from various chatbot domains.
+For more info on the CLINC150 dataset, check out the
+[dataset repo](https://github.com/clinc/oos-eval).
+
+Firstly, let's clone the dataset locally:
+```bash
+git clone https://github.com/clinc/oos-eval.git
```
Comment (tadejsv) [inline docs/get-started/clinc150/index.md:41]: I think you can provide a link to only the datasets needed to download with `curl`

File: docs/get-started/clinc150/index.md [hunk 10]
```diff
@@ -0,0 +1,843 @@
...
+simple nearest neighbor rule.
+
+
+## CLINC150
+
+We will use the CLINC150 dataset as the base of our experiment. It is a dataset of
+utterance-intent pairs and is commonly used for evaluating intent models. It comes
+in train, val and test splits and contains 150 intents from various chatbot domains.
+For more info on the CLINC150 dataset, check out the
+[dataset repo](https://github.com/clinc/oos-eval).
```
Comment (tadejsv) [inline docs/get-started/clinc150/index.md:37]: Maybe show a few examples of entries in the dataset here

File: docs/get-started/clinc150/index.md [hunk 11]
```diff
@@ -0,0 +1,843 @@
...
+{
+  "r_precision": 0.6597606769142386,
+  "precision_at_k": 0.5920919540229873,
+  "recall_at_k": 0.5920919540229873,
+  "f1_score_at_k": 0.5920919540229873,
+  "average_precision": 0.817989559018391,
+  "hit_at_k": 0.9962222222222222,
+  "reciprocal_rank": 0.9344587442949851,
+  "dcg_at_k": 6.55171986283775,
+  "ndcg_at_k": 0.908690691171733
```
Comment (tadejsv) [inline docs/get-started/clinc150/index.md:225]: I would only show 2-3 decimals here

File: docs/get-started/clinc150/index.md [hunk 12]
```diff
@@ -0,0 +1,843 @@
...
+{
+  "r_precision": 0.9152863975803448,
+  "precision_at_k": 0.8819923371647765,
+  "recall_at_k": 0.8819923371647765,
+  "f1_score_at_k": 0.8819923371647765,
+  "average_precision": 0.9500754908924715,
+  "hit_at_k": 0.9917777777777778,
+  "reciprocal_rank": 0.9711110828700991,
+  "dcg_at_k": 8.840575660386225,
+  "ndcg_at_k": 0.9684599889599816
```
Comment (tadejsv) [inline docs/get-started/clinc150/index.md:358]: I would only show 2-3 decimals here

File: docs/get-started/clinc150/index.md [hunk 13]
```diff
@@ -0,0 +1,843 @@
...
+print(f'Avg num utterances per intent in test set: {num_utts_per_intent_test}')
+```
+```
+Num train intents: 150
+Num val intents: 150
+Num test intents: 150
+Avg num utterances per intent in train set: 100.0
+Avg num utterances per intent in val set: 20.0
+Avg num utterances per intent in test set: 30.0
+```
```
Comment (tadejsv) [inline docs/get-started/clinc150/index.md:121]: As this tutorial is quite long, I would omit this code part entirely, and just state these numbers in text - especially since they are "fixed" and already mentioned at the beginning of dataset readme

File: docs/get-started/clinc150/index.md [hunk 14]
```diff
@@ -0,0 +1,843 @@
...
+
+As described above, we will use `finetuner` to finetune an embedding model in order
+to bring representations of the same intent, closer in the embedding space. For that
+we will use the `transformers` library to define a transformer-based embedding model.
+We will load a pre-trained transformer as our starting point, the `paraphrase-MiniLM-L6-v2`
+model from `sentence-transformers`. Check out the links below for more information on
+this specific model as well as sentence transformers in general:
+
+* <https://huggingface.co/sentence-transformers/paraphrase-MiniLM-L6-v2>
+* <https://www.sbert.net/index.html>
```
Comment (tadejsv) [inline docs/get-started/clinc150/index.md:133]: I would ommit this, and just make clickable links when you refer to the model and sentence-transformers

File: docs/get-started/clinc150/index.md [hunk 15]
```diff
@@ -0,0 +1,843 @@
...
+        return mean_pooling(out, inputs['attention_mask'])
+```
+
+Our model is a pre-trained embedding model, i.e. it outputs a high-dimensional representation
+given some input text and it has been pre-trained on large text corpora. To use the model
+defined above, we need to be able to convert raw text to the tensor format that our model
+accepts as input. To do that, we need to use the BPE tokenizer, provided by the `transformers`
+package, that converts texts to BPE encoded arrays that our model accepts as input. We make
+use of the collate function that `finetuner` supports. The collate function offers a way to
+specify the conversion of batch elements to model input tensors.
```
Comment (tadejsv) [inline docs/get-started/clinc150/index.md:166]: Split this paragraph into two

**Review by tadejsv — CHANGES_REQUESTED** [review]

**Review by azayz — APPROVED** [review]
LGTM

File: docs/get-started/clinc150/index.md (hunk 13 above)
Comment (gmastrapas) [reply docs/get-started/clinc150/index.md:121]: there are variables in this code block that are used afterwards

Comment (tadejsv) [reply docs/get-started/clinc150/index.md:121]: But those variables are already known (stated in readme of dataset)

File: docs/get-started/clinc150/index.md [hunk 16]
```diff
@@ -0,0 +1,843 @@
...
+```
+```
+Evaluating FINE-TUNED model on test data:
+{
+  "r_precision": 0.9152863975803448,
+  "precision_at_k": 0.8819923371647765,
+  "recall_at_k": 0.8819923371647765,
+  "f1_score_at_k": 0.8819923371647765,
+  "average_precision": 0.9500754908924715,
+  "hit_at_k": 0.9917777777777778,
```
Comment (bwanglzu) [inline docs/get-started/clinc150/index.md:355]: i would omit the evaluation print, but turn the results into a table (pretrained vs finetuned), and as Tadej said, use 2 decimals.

**Review by bwanglzu — DISMISSED** [review]
minor comments, my general impression is it's a bit complicated (in term of length)

File: docs/get-started/clinc150/index.md (hunk 16 above)
Comment (gmastrapas) [reply docs/get-started/clinc150/index.md:355]: you mean keep the code that prints, but instead of showing the console output display everything in a table?

**Review by tadejsv — APPROVED** [review]
LGTM!

**Review by bwanglzu — COMMENTED** [review]
LGTM!
